### PR TITLE
Updated Initialization for Style Transfer

### PR DESCRIPTION
**Pull request by abhishekpratapa**
# Overview
Initialization for style transfer is a uniform between -0.7 and 0.7. Previously only pertained weights were used for this in TensorFlow.

# Fixes
- Weight initialization in style transfer to use uniform between -0.7 and 0.7
- Random Seed values for initialization

### CODE REVIEW

File: src/toolkits/style_transfer/style_transfer_model_definition.cpp [hunk 1]
```diff
@@ -20,14 +22,24 @@ using CoreML::Specification::NeuralNetworkLayer;
...
+constexpr float UPPER_BOUND = 0.7; 
+
 // TODO: refactor code to be more readable with loops
-void define_resnet(model_spec& nn_spec, size_t num_styles) {
+void define_resnet(model_spec& nn_spec, size_t num_styles, int random_seed=0) {
+  std::mt19937 random_engine;
+  std::seed_seq seed_seq{random_seed};
+  random_engine = std::mt19937(seed_seq);
+
+  auto initializer = xavier_weight_initializer(LOWER_BOUND, UPPER_BOUND, &random_engine);
```
Comment (TobyRoseman) [inline src/toolkits/style_transfer/style_transfer_model_definition.cpp]: I'm really confused by this line. First off, it would probably be best not to use `auto` here.

This is what you're calling, right? https://github.com/apple/turicreate/blob/fb1725cffa5cbd188e281626771cb8b6a53b5ae8/src/ml/neural_net/weight_init.cpp#L25

If so, that doesn't look right to me. The first two parameters seem wrong. fan in/out seems different than lower/upper bound. At the very least these are getting covered from float to size_t which does seem right since they aren't whole numbers.

File: src/toolkits/style_transfer/style_transfer.cpp [hunk 2]
```diff
@@ -686,6 +686,12 @@ void style_transfer::init_train(gl_sarray style, gl_sarray content,
   }
   size_t num_styles = num_styles_iter->second;
 
+  auto pretrained_weights_iter = opts.find("pretrained_weights");
```
Comment (TobyRoseman) [inline src/toolkits/style_transfer/style_transfer.cpp]: Is `auto` helping or hurting readability here?

File: src/toolkits/style_transfer/style_transfer_model_definition.cpp (hunk 1 above)
Comment (abhishekpratapa) [reply src/toolkits/style_transfer/style_transfer_model_definition.cpp]: Good catch! I was testing it and the stylization didn't work, properly.

**Comment by nickjong** [comment]
What is the relationship between this PR and #2874 ?

File: src/ml/neural_net/weight_init.hpp [hunk 3]
```diff
@@ -48,6 +48,32 @@ class xavier_weight_initializer {
...
 };
 
+
+class uniform_weight_initializer {
+  public:
+
+  /**
+   * Creates a weight initializer that performs Uniform initialization
+   *
+   * \param lower_bound The number of inputs that affect each output from the layer
```
Comment (nickjong) [inline src/ml/neural_net/weight_init.hpp]: This documentation does not appear to be correct

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py [hunk 4]
```diff
@@ -247,6 +247,7 @@ def create(
     options["num_styles"] = len(style_dataset)
     options["resnet_mlmodel_path"] = pretrained_resnet_model.get_model_path("coreml")
     options["vgg_mlmodel_path"] = pretrained_vgg16_model.get_model_path("coreml")
+    options["pretrained_weights"] = params["pretrained_weights"]
```
Comment (nickjong) [inline src/python/turicreate/toolkits/style_transfer/style_transfer.py]: Why is this needed?

File: src/toolkits/style_transfer/style_transfer.cpp (hunk 2 above)
Comment (nickjong) [reply src/toolkits/style_transfer/style_transfer.cpp]: I believe the alternative is
```
std::map<std::string, flexible_type>::iterator pretrained_weights_iter = opts.find("pretrained_weights");
```
So I would argue yes.

File: src/toolkits/style_transfer/style_transfer_model_definition.cpp [hunk 5]
```diff
@@ -20,14 +22,24 @@ using CoreML::Specification::NeuralNetworkLayer;
...
+constexpr float UPPER_BOUND = 0.07; 
+
 // TODO: refactor code to be more readable with loops
-void define_resnet(model_spec& nn_spec, size_t num_styles) {
+void define_resnet(model_spec& nn_spec, size_t num_styles, int random_seed=0) {
+  std::mt19937 random_engine;
+  std::seed_seq seed_seq{random_seed};
+  random_engine = std::mt19937(seed_seq);
+
+  auto initializer = uniform_weight_initializer(LOWER_BOUND, UPPER_BOUND, &random_engine);
```
Comment (nickjong) [inline src/toolkits/style_transfer/style_transfer_model_definition.cpp]: This one I think is more readable, well idiomatic, as
```
uniform_weight_initializer initializer(LOWER_BOUND, UPPER_BOUND, &random_engine);
```

File: src/toolkits/style_transfer/style_transfer_model_definition.cpp [hunk 6]
```diff
@@ -46,7 +58,7 @@ void define_resnet(model_spec& nn_spec, size_t num_styles) {
       /* stride_height */ 1,
       /* stride_width */ 1,
       /* padding */ padding_type::VALID,
-      /* weight_init_fn */ zero_weight_initializer());
+      /* weight_init_fn */ initializer);
```
Comment (nickjong) [inline src/toolkits/style_transfer/style_transfer_model_definition.cpp]: Doesn't this mean we're performing a bunch of random initialization even when we're about to overwrite with the pre-trained model?

**Comment by abhishekpratapa** [comment]
@nickjong 
> What is the relationship between this PR and #2874 ?

This fixes the waviness users were experiencing. The pre-trained model let's the users get faster models but this is more likely to introduce artifacts into the stylization. The uniform initialization let's the user train a higher quality model but with a higher number of iterations. It resolves (7.) of the issue

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py (hunk 4 above)
Comment (abhishekpratapa) [reply src/python/turicreate/toolkits/style_transfer/style_transfer.py]: This is to allow the user to specify through the advanced parameters whether they want uniform initialization or initialization from the pre-trained model.

File: src/toolkits/style_transfer/style_transfer_model_definition.cpp (hunk 6 above)
Comment (abhishekpratapa) [reply src/toolkits/style_transfer/style_transfer_model_definition.cpp]: Yeah, I noticed that as well. I wasn't sure if it was worth it to write in a flag to keep the model uninitialized. It seemed like a small one time startup cost for the model, but I'll add a flag [Boolean] (if you think that's the right way to do it). Or should it be an enum type of initialization for future proofing? What's your suggestion?

Comment (abhishekpratapa) [reply src/toolkits/style_transfer/style_transfer_model_definition.cpp]: I think I got it.

File: src/toolkits/style_transfer/style_transfer_model_definition.cpp [hunk 7]
```diff
@@ -33,12 +35,20 @@ constexpr float LOWER_BOUND = -0.07;
...
+void define_resnet(model_spec& nn_spec, size_t num_styles, bool initialize=false, int random_seed=0) {
+  weight_initializer initializer;
+
+  // This is to make sure that when the uniform initialization is not needed extra work is avoided
+  if (initialize) {
+    std::mt19937 random_engine;
+    std::seed_seq seed_seq{random_seed};
+    random_engine = std::mt19937(seed_seq);
+
+    initializer = uniform_weight_initializer(LOWER_BOUND, UPPER_BOUND, &random_engine);
```
Comment (nickjong) [inline src/toolkits/style_transfer/style_transfer_model_definition.cpp]: I believe `&random_engine` will become a dangling pointer once `random_engine` goes out scope after this line. I suspect you will need to move `random_engine` outside the scope of this `if` statement to obtain well-defined behavior

Comment (shreyajain17) [reply src/toolkits/style_transfer/style_transfer_model_definition.cpp]: > @nickjong
> 
> > What is the relationship between this PR and #2874 ?
> 
> This fixes the waviness users were experiencing. The pre-trained model let's the users get faster models but this is more likely to introduce artifacts into the stylization. The uniform initialization let's the user train a higher quality model but with a higher number of iterations. It resolves (7.) of the issue

Are we exposing both pre trained weights as well as just the uniform initialization. I think if we are exposing it would be good for the users to actually have this literature somewhere. I don't think it is intuitive and I think we should put it in our userguide. 

I also think maybe in another PR you should expose the random_seed parameter. I think it would be great if users could generate reproducible models.

Comment (nickjong) [reply src/toolkits/style_transfer/style_transfer_model_definition.cpp]: I propose trying to have all the deep-learning toolkits controlled by random seeds for 6.2...
